src/common/utils/fold_test.cc

```diff
@@ -0,0 +1,96 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "src/common/utils/fold.h"
+
+#include "absl/strings/str_join.h"
+#include "src/common/testing/gtest.h"
+
+namespace raksha::common::utils {
+
+struct IntVecAndSum {
+  std::vector<int> int_vec;
+  int sum;
+};
+
+class SumFoldTest : public testing::TestWithParam<IntVecAndSum> {};
+
+TEST_P(SumFoldTest, SumFoldTest) {
+  const auto &[vec, expected_result] = GetParam();
+  auto sum_fn = [](int sum_so_far, int num) { return sum_so_far + num; };
+  EXPECT_EQ(fold(vec, 0, sum_fn), expected_result);
+  EXPECT_EQ(rfold(vec, 0, sum_fn), expected_result);
+}
+
+static const IntVecAndSum kIntVecsAndSum[] = {
+    {.int_vec = {}, .sum = 0},
+    {.int_vec = {1}, .sum = 1},
+    {.int_vec = {1, 2, 3, 4}, .sum = 10},
+    {.int_vec = {-1, 5, 2, -8}, .sum = -2}};
+
+INSTANTIATE_TEST_SUITE_P(SumFoldTest, SumFoldTest,
+                         testing::ValuesIn(kIntVecsAndSum));
+
+class ConcatFoldTest : public testing::TestWithParam<std::vector<std::string>> {
+};
+
+TEST_P(ConcatFoldTest, ConcatFoldTest) {
+  const std::vector<std::string> &string_vec = GetParam();
+  auto concat_fn = [](std::string concat_so_far, const std::string &input) {
+    // Yes, this is bad form in general, but the quadraticness shouldn't matter
+    // for a test.
+    return concat_so_far + input;
+  };
+  EXPECT_EQ(fold(string_vec, std::string(""), concat_fn),
+            absl::StrJoin(string_vec, ""));
```

Comment:
Instead of computing `StrJoin`, it would be more clear to pass the expected string itself (just like `IntVecSum` case)? 

(I don't feel strongly about this specific test. However, in general, I think it is better to avoid computation of expected values because we could inadvertently introduce bugs in expected values and have tests pass when they shouldn't.)

Reply:
Done.